src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -0,0 +1,703 @@
+    };
+    for (const borrower_t& b : borrowers) {
+        const ingress_egress_state<quota_t> share{
+          .in = equal_share_or_0(split.in, b.borrowers_count.in),
+          .eg = equal_share_or_0(split.eg, b.borrowers_count.eg)};
+        if (!is_zero(share)) {
+            // TBD: make the invocation parallel
+            co_await container().invoke_on(
+              b.shard_id,
+              [share](snc_quota_manager& qm) { qm.adjust_quota(share); });
+        }
+    }
+    if (split.in.rem != 0 || split.eg.rem != 0) {
+        vlog(klog.error, "qb - Split not dispensed in full: {}", split);
+    }
+}
+
+namespace {
+
+/// Split \p value between the elements of vector \p target in full, adding them
+/// to the elements already in the vector.
+/// If \p value is not a multiple of target.size(), the entire \p value
+/// is still dispensed in full. Quotinents rounded towards infinity
+/// are added to the front elements of the \p target, and those rounded
+/// towards zero are added to the back elements.
+void dispense_equally(std::vector<quota_t>& target, const quota_t value) {
+    auto share = std::div(value, target.size());
+    for (quota_t& v : target) {
+        v += share.quot;
+        if (share.rem > 0) {
+            v += 1;
+            share.quot -= 1;
+        } else if (share.rem < 0) {
+            v -= 1;
+            share.quot += 1;
+        }
+    }
+}
+
+/// If \p value is less than \p limit, set it to the \p limit value and
+/// return the difference as positive number. Otherwise \p value is unchanged
+/// and 0 is returned
+quota_t cap_to_ceiling(quota_t& value, const quota_t limit) {
+    if (const quota_t d = limit - value; d > 0) {
+        value += d;
+        return d;
+    }
+    return 0;
+}
+
+/// If \p delta is negative, splits its value accourding to \p surplus values:
+/// - amount under total surplus is split pro rata to \p surplus amounts
+/// - amount over total surplus is split equally. Values in \p schedule
+/// are added to.
+/// \pre schedule.size() == surplus.size()
+void dispense_negative_deltas(
+  std::vector<quota_t>& schedule,
+  quota_t& delta,
+  const std::vector<quota_t>& surplus) {
+    if (delta >= 0) {
+        return;
+    }
+
+    const quota_t total_surplus = std::reduce(
+      surplus.cbegin(), surplus.cend(), quota_t{0}, std::plus{});
+
+    if (delta > -total_surplus) {
+        quota_t remainder = delta;
+        // pro rata to surpluses
+        for (size_t k = 0; k != ss::smp::count; ++k) {
+            const quota_t share = muldiv(delta, surplus.at(k), total_surplus);
+            schedule.at(k) += share;
+            remainder -= share;
+        }
+        // the remainder equally
+        if (remainder > 0) {
+            vlog(
+              klog.error,
+              "qb - Expected {} <= 0; delta: {}, total_surplus: {}, surpluses: "
+              "{}",
+              remainder,
+              delta,
+              total_surplus,
+              surplus);
+        }
```

Comment:
would this be a case where you'd want to return early instead of carrying on, since before IIRC it was a vassert(false) situation?

EDIT: oh i see that the effect of the remaining conditional is that the error condition becomes a noop. consider adding a `return` after the vlog to drive that point across.